Experiment automation/(LUMICKS) Reference kit automation/referencekit_automation.py

```diff
@@ -0,0 +1,335 @@
+                if get_force() > force_threshold:
+                    print("Looks like only one tether is left, continue with the rest of the protocol")
+                    one_tether_remains = True
+                else:
+                    raise RuntimeError("Lost all tethers")
+            n += 1
+
+
+def fd_workflow(experiment_name, path, match_threshold, dna_fishing_speed, min_distance_fishing, max_distance_fishing,
+                force_threshold, tether_lost_threshold, max_kymos, force_kymo, dna_length):
+
+    """Main workflow of the script. Catches beads, fishes for DNA and makes a kymograph at 10 pN.
+    """
+    path = validate_dir(path)
+
+    kymo_count = -1
+
+    while kymo_count < max_kymos:
+        try:
+            kymo_count += 1
+            catch_beads(match_threshold=match_threshold)
+            catch_dna(min_distance=min_distance_fishing, max_distance=max_distance_fishing,
+                      fishing_speed=dna_fishing_speed, match_threshold=match_threshold, force_threshold=force_threshold,
+                      fishing_attempts=10, dt=0.5, dna_length = dna_length)
+            stop_flow()
+            bl.pause(3.0)
+            bl.microstage.move_to(name_buffer_channel)
+            check_multiple_tethers(max_distance=max_distance_fishing, multiple_tether_force=10, force_threshold=force_threshold,
+                                   match_threshold=match_threshold, max_retries=7, speed=dna_fishing_speed, dna_length=dna_length)
+            print("Go to distance of 0.75 x tether length")
+            goto_distance(0.75 * dna_length, match_threshold, speed=dna_fishing_speed)
+            bl.pause(1.0)
+            print("reset force")
+            bl.reset_force()
+            print("Go to distance of 1 x tether length")
+            goto_distance(dna_length, match_threshold, speed=dna_fishing_speed)
+            check_tether_breakage(tether_lost_threshold)
+            print(f"go to a force of {force_kymo} pN")
+            goto_force(force_kymo, match_threshold)
+            print("Go to channel junction")
+            bl.microstage.move_to(name_junction, speed=1)  # speed in mm/s
+            bl.pause(2.0)
+            print("Go to protein channel")
+            bl.microstage.move_to(name_protein_channel, speed=1)  # speed in mm/s
+            bl.pause(2.0)
+            print(f"Recording kymograph (kymo: {kymo_count}).")
+            current_time = datetime.now().strftime("%H%M%S")
+            name = f"{current_time}-{experiment_name}_tether={kymo_count}"
+            check_tether_breakage(tether_lost_threshold)
+            make_kymograph(force_kymo, match_threshold, name, path)
+            print("Script finished!")
+
+        except RuntimeError as e:
+            if kymo_count == max_kymos:
+                print(f"Reached max number of {max_kymos} attempts, stop script")
+            else:
+                print("Restarting protocol.")

```

Comment:
This logic isn't really needed if you just use a for loop.

Reply:
That's fair. One thing you could do is only fix up the parts that you are comfortable with changing then (doc-strings and whatnot). Then in a separate PR do the other changes, but just leave that up for the time being until there is a time to test. It's not super critical, just a bit easier to read / maintain.

Reply:
That sounds like a good plan. Maybe there will also be feedback from users that we can then incorporate at the same time.